processor/src/chiplets/hasher/mod.rs

```diff
@@ -252,78 +255,64 @@ impl Hasher {
+                // permutations need to be set as follows:
+                // - first permutation: init linear hash on the first row, and absorb the next
+                //   operation batch on the last row.
+                // - middle permutations: continue hashing on the first row, and absorb the next
+                //   operation batch on the last row.
+                // - last permutation: continue hashing on the first row, and return the result
+                //   on the last row.
                 self.trace.append_permutation(&mut state, START, ABSORB);
-            }
 
-            let mut last_state = state;
-
-            for batch in op_batches.iter().take(num_batches - 1).skip(1) {
-                absorb_into_state(&mut state, batch.groups());
-                // add the lookup for absorbing the next operation batch.
-                self.append_lookup(
-                    ABSORB_LABEL,
-                    last_state,
-                    ZERO,
-                    HasherLookupContext::Absorb(state),
-                );
-
-                if is_memoized {
-                    self.trace
-                        .copy_trace(&mut state, row..(row + HASH_CYCLE_LEN));
-                    row += HASH_CYCLE_LEN;
-                } else {
+                for batch in op_batches.iter().take(num_batches - 1).skip(1) {
+                    absorb_into_state(&mut state, batch.groups());
+                    // add the lookup for absorbing the next operation batch.
+                    self.append_lookup(ABSORB_LABEL, ZERO, HasherLookupContext::Absorb);
                     self.trace.append_permutation(&mut state, CONTINUE, ABSORB);
                 }
 
-                last_state = state;
+                absorb_into_state(&mut state, op_batches[num_batches - 1].groups());
+                // add the lookup for absorbing the final operation batch.
+                self.append_lookup(ABSORB_LABEL, ZERO, HasherLookupContext::Absorb);
+                self.trace.append_permutation(&mut state, CONTINUE, RETURN);
             }
-
-            absorb_into_state(&mut state, op_batches[num_batches - 1].groups());
-            // add the lookup for absorbing the final operation batch.
-            self.append_lookup(
-                ABSORB_LABEL,
-                last_state,
-                ZERO,
-                HasherLookupContext::Absorb(state),
-            );
-            if is_memoized {
-                self.trace.copy_trace(&mut state, row..end_row);
+            self.insert_to_memoized_trace_map(addr, expected_hash);
+        } else {
+            // let mut row = start_row;
+            if num_batches == 1 {
+                // copy trace if trace exists
+                self.trace.copy_trace(&mut state, start_row..end_row);
             } else {
-                self.trace.append_permutation(&mut state, CONTINUE, RETURN);
+                for (i, _) in op_batches.iter().take(num_batches - 1).skip(1).enumerate() {
```

Comment:
Since this doesn't actually use the `op_batch` data I would simplify it to just iterate over the desired range, e.g. 

```
for i in 1..num_batches - 1 { ... }
```